### PR TITLE
[DependencyInjection] Use a service locator in AddConstraintValidatorsPass

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

Use a service locator to load constraint validators: it allows them to be private.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 namespace Symfony\Bundle\FrameworkBundle\Validator;
 
-use Symfony\Component\DependencyInjection\ContainerInterface;
+use Psr\Container\ContainerInterface;
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: Given $container is protected I think that's a BC break

Comment (GuilhemN) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: I think it's acceptable, the methods defined are the same and it will only fail if someone checks its interface which is unlikely.

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: Not sure. ContainerInterface contains more methods than the Psr one, calling `$this->container->getParameter()` would break. We had to write a BC layer for a similar case in #21451

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: Note also that adding deprecations for using experimental features has been rejected in #21625.

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: about deprecating in this PR, see https://github.com/symfony/symfony/pull/21625#issuecomment-281973526
this needs to be resolved before acting to me.

Comment (GuilhemN) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: I've an idea of a BC layer (including deprecations but not about experimental features), I'll update the pr soon.

**Comment by GuilhemN** [comment]
Considering your comment, I think it's better to delay this change to 4.0 and to just deprecate `ConstraintValidatorFactory::$container` and `ConstraintValidatorFactory::$validators` for now, it will be much easier to do this way.

**Comment by nicolas-grekas** [comment]
Is there any BC break really?
I mean, code that exists won't suddently get a PSR-11 only object injected, will they?

**Comment by GuilhemN** [comment]
If someone uses `$container` he would receive a psr-11 only container as I updated the first argument to a service locator.
I think we can change the type hint now but we can't update the service config to a service locator before 4.0.

**Comment by nicolas-grekas** [comment]
> If someone uses $container he would receive a psr-11 only container as I updated the first argument to a service locator

BC breaks are about existing code - ie already written ones - that may break because of a change here. How can this change break existing code?

**Comment by chalasr** [comment]
Anyone extending this code and wiring it on its own doesn't have to check that what it receives is a symfony container until now since we guarantee it, after this change such code would work by luck if it relies on any method that is not part of the PSR-11 replacement. It would truly break only when decorating the service of this factory, maybe we don't care.

**Comment by GuilhemN** [comment]
@chalasr as discussed with @nicolas-grekas, there is no bc break as it won't break existing code: the only time the container will no longer be an instance of the symfony container when upgrading will be when using the `validator.validator_factory` service but it doesn't matter as we know it's not extended (its class is fixed).
So maybe a class extending `ConstraintValidatorFactory` won't support the psr-11 container but that's not truly a bc break as it won't break existing code, the container will always be a symfony container, only new code could be broken but that's not part of bc.

**Comment by chalasr** [comment]
Pleased to see one more container injection removed then.

File: src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php [hunk 2]
```diff
@@ -70,11 +72,15 @@ public function getInstance(Constraint $constraint)
...
+                $this->validators[$name] = $this->container->get($name);
+            } else {
+                if (!class_exists($name)) {
+                    throw new ValidatorException(sprintf('Constraint validator "%s" does not exist or it is not enabled. Check the "validatedBy" method in your constraint class "%s".', $name, get_class($constraint)));
+                }
 
-            $this->validators[$name] = new $name();
+                $this->validators[$name] = new $name();
+            }
         } elseif (is_string($this->validators[$name])) {
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: should `has()` be called here too?

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php (hunk 2 above)
Comment (GuilhemN) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: Why? In case what is returned by the container is a string ?

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: `get()` is called if this check succeeds. We're checking `has()` before `get()` in the first check above so I'm wondering if we need it here

Comment (GuilhemN) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: Above it's needed because we want to know when to fallback to manual instantiating, here it just depends of the exception we want, with a `has` the code will fail a few lines later with an InvalidArgumentException, otherwise it will fail in the container.
This should not happen unless the user entered a wrong service, imo it's not worth supporting it.

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Validator/ConstraintValidatorFactory.php]: 👍

**Comment by chalasr** [comment]
The Validator's CHANGELOG should mention the deprecation

**Comment by GuilhemN** [comment]
Changelog updated ☺

File: UPGRADE-3.3.md [hunk 3]
```diff
@@ -71,6 +71,9 @@ FrameworkBundle
    deprecated and will be removed in 4.0. Use the `Symfony\Component\Form\DependencyInjection\FormPass`
    class instead.
 
+ * Extending `ConstraintValidatorFactory` is deprecated and won't be supported
+   in 4.0.
```
Comment (nicolas-grekas) [inline UPGRADE-3.3.md]: could be on one line

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php [hunk 4]
```diff
@@ -26,22 +27,13 @@ public function process(ContainerBuilder $container)
...
+                $validators[$attributes[0]['alias']] = new Reference($id);
             }
 
             $definition = $container->getDefinition($id);
-
-            if (!$definition->isPublic()) {
-                throw new InvalidArgumentException(sprintf('The service "%s" must be public as it can be lazy-loaded.', $id));
-            }
-
-            if ($definition->isAbstract()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php]: Abstract services should be ignored ie you should "continue". Note that this applies to many other cases in the code base: many passes throw on abstract services, but now that tags can be inherited in child definitions, we should just "continue" insteand. Deserves another PR if you'd like to do it :)

File: UPGRADE-3.3.md (hunk 3 above)
Comment (GuilhemN) [inline UPGRADE-3.3.md]: updated

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php (hunk 4 above)
Comment (GuilhemN) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php]: Seems like `ContainerBuilder::findTaggedServiceIds()` deserves a new argument ;) 
I can't do it this week so I opened https://github.com/symfony/symfony/issues/21761 in case someone is interested.

**Comment by nicolas-grekas** [comment]
rebase needed

**Comment by GuilhemN** [comment]
@nicolas-grekas done.

Also, abstract definitions are now ignored and I improved the test readability by removing mocks.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by fabpot** [comment]
Thank you @GuilhemN.

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/AddConstraintValidatorsPassTest.php [hunk 5]
```diff
@@ -13,56 +13,34 @@
...
-        $validatorFactoryDefinition->expects($this->once())
-            ->method('replaceArgument')
-            ->with(1, array(
-                'My\Fully\Qualified\Class\Named\Validator1' => 'my_constraint_validator_service1',
-                'my_constraint_validator_alias1' => 'my_constraint_validator_service1',
-                'My\Fully\Qualified\Class\Named\Validator2' => 'my_constraint_validator_service2',
-            ));
+        $container = new ContainerBuilder();
+        $validatorFactory = $container->register('validator.validator_factory')
+            ->setArguments(array(new ServiceLocatorArgument()));
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/AddConstraintValidatorsPassTest.php]: broken here, the constructor expects an argument. See https://github.com/symfony/symfony/pull/21821
